sklearn/preprocessing/tests/test_discretization.py

```diff
@@ -35,6 +50,20 @@ def test_valid_n_bins():
     assert KBinsDiscretizer(n_bins=2).fit(X).n_bins_.dtype == np.dtype(int)
 
 
+def test_invalid_sample_weight():
+    # sample_weight parameter is used with wrong strategy (other than quantile)
+    strategy = ["uniform", "kmeans"]
+    sample_weight = [1, 1, 1, 1]
+    for s in strategy:
+        est = KBinsDiscretizer(n_bins=3, strategy=s)
+        err_msg = (
+            "`sample_weight` was provided but it can be only used with"
+            f"strategy='quantile'. Got strategy={s!r} instead."
+        )
+        with pytest.raises(ValueError, match=err_msg):
+            est.fit_transform(X, sample_weight=sample_weight)
```

Comment:
```suggestion
    sample_weight = np.ones(size=(X.shape[0]))
    estimator = KBinsDiscretizer(n_bins=3, strategy=strategy)
    err_msg = (
        "`sample_weight` was provided but it can be only used with"
        "strategy='quantile'."
    )
    with pytest.raises(ValueError, match=err_msg):
        est.fit(X, sample_weight=sample_weight)
```

Reply:
At line 55 (the suggestion in green): ```X.shape[0]``` doesn't work. ```X``` is a list, it doesn't have attribute ```shape```.  Maybe ```size=(len(X))```? Unless it should be converted to an array.

Reply:
On line 55 ```sample_weight = np.ones(size=(X.shape[0]))```: 
X is a list, so we can't use the attribute shape. 
Also ```np.ones``` has a parameter  ```shape```, not ```size```. 
Maybe then ```sample_weight = np.ones(shape=len(X))```?

Reply:
Yes it is indeed the first line of my suggestion :)